crates/sui-framework/packages/move-stdlib/tests/integer_tests.move

```diff
@@ -104,6 +106,67 @@ public(package) macro fun test_divide_and_round_up($max: _, $cases: vector<_>) {
     })
 }
 
+public(package) macro fun check_mul_div($max: _, $x: _, $y: _, $z: _) {
+    let x = $x;
+    let y = $y;
+    let z = $z;
+    let max = $max;
+    if (z == 1) {
+        assert_eq!(x.mul_div(y, z), x.mul(y));
+        assert_eq!(x.mul_div_ceil(y, z), x.mul(y));
+    };
+
+    if (x == 0 || y == 0) {
+        assert_eq!(x.mul_div(y, z), 0);
+        assert_eq!(y.mul_div(x, z), 0);
+        assert_eq!(x.mul_div_ceil(y, z), 0);
+        assert_eq!(y.mul_div_ceil(x, z), 0);
+    };
+
+    if ((x == max && y < z) || (y == max && x < z)) {
+        assert_lt!(x.mul_div(y, z), max);
+        assert_lt!(y.mul_div(x, z), max);
+    }; // TODO: Test abort_overflow
+
+    if (x <= max.div(y) || y <= max.div(x)) {
+        assert_eq!(x.mul_div(y, z), (x.mul(y)) / z);
+        assert_eq!(y.mul_div(x, z), (y.mul(x)) / z);
+    };
+
+    if (x % z == 0 || y % z == 0) {
+        assert_eq!(x.mul_div_ceil(y, z), x.mul_div(y, z));
+    };
+
+    if (x * y < z) {
+        assert_eq!(x.mul_div(y, z), 0);
+        assert_eq!(x.mul_div_ceil(y, z), 1);
+    };
+
+    assert_le!(x.mul_div(y, z), max);
+    assert_le!(y.mul_div(x, z), max);
+    assert_le!(x.mul_div_ceil(y, z) - x.mul_div(y, z), 1);
+    assert_le!(y.mul_div_ceil(x, z) - y.mul_div(x, z), 1);
+}
+
+public(package) macro fun test_mul_div($max: _, $cases: vector<_>) {
+    let max = $max;
+    let cases = $cases;
+    assert_eq!(max.mul_div(max, max.max(1)), max);
+    cases!(max, cases, |_case_pred, case, _case_succ| {
+        assert_eq!(case.mul_div(case, case.max(1)), case); // avoid division by 0
+        // TODO: more test cases?
+    });
+    check_mul_div!(max, 6, 4, 3);
+    check_mul_div!(max, 7, 5, 1);
+    check_mul_div!(max, 5, 3, 7);
+    check_mul_div!(max, 5, 3, 20);
+    check_mul_div!(max, 0, 10, 5);
+    check_mul_div!(max, max, 2, 2);
+    check_mul_div!(max, max, 2, 10);
+    check_mul_div!(max, max/2+10, max/2+15, max);
+    check_mul_div!(max, max, max, max);
```

Comment:
I would prefer to see the cases be used in some way instead of these hard coded ones (thats the point of cases)